internal/component/loki/source/docker/internal/dockertarget/target.go

```diff
@@ -235,9 +235,11 @@ func (t *Target) StartIfNotRunning() {
 
 // Stop shuts down the target.
 func (t *Target) Stop() {
-	t.cancel()
-	t.wg.Wait()
-	level.Debug(t.logger).Log("msg", "stopped Docker target", "container", t.containerName)
+	if t.Ready() {
+		t.cancel()
+		t.wg.Wait()
```

Comment:
Isn't the nil pointer dereference due to a `t` which is nil? I thought all we need to do is:
```
if t == nil {
  return
}
```

I would prefer if we don't call `t.Ready()` without mutexes, since the state of the container might change between the call to `t.Ready()` and the call to stop. I know it's unlikely, but it feels more resilient.

Reply:
I think that it's actually the cancel function which is nil in this case. if you check the logs in the issue https://github.com/grafana/alloy/issues/868, there is an error line right before the panic with the id of the target returned by t.target.Name().
If the state changes in between that would not be a problem because if the target ran once, there should be a cancel function associated with it. So in the very specific case where ready is true and the process loop ends setting it to false right before the cancel call, then the cancel would be a no-op